core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java

```diff
@@ -104,13 +114,13 @@ public static LoggingServiceBuilder builder() {
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
         if (sampler.isSampled(ctx)) {
-            ctx.log().addListener(log -> logRequest(((ServiceRequestContext) log.context()).logger(),
-                                                    log,
+            final Logger logger = firstNonNull(this.logger, ctx.logger());
```

Comment:
We do not want `ctx.logger()` to be invoked when `this.logger` is non-null:

```java
final Logger logger = this.logger != null ? this.logger : ctx.logger();
```

Reply:
If what I know correct,
```java
final Logger logger = firstNonNull(this.logger, ctx.logger());
```
works similar to the below.
```java
final Logger logger = this.logger != null ? this.logger : ctx.logger();
```

The only difference is that `firstNonNull()`explicitly throws a `NullPointException` if both are null.
May I use `firstNonNull()`?
(https://github.com/line/armeria/pull/2237#discussion_r350063812 too)

I'll fix it after the conversation.  🙇‍♂️

Reply:
IIUC, The main difference is that `firstNonNull(this.logger, ctx.logger());` evaluates `ctx.logger()`.
if we use 3 way operator we can't avoid to call `ctx.logger()` if `this.logger` is non-null.

Sorry for the confusion. I reviewed it in the wrong direction.

Reply:
@ikhoon 
Oh.. evaluation time is different! 🙇‍♂️

Reply:
fixed